src/main/java/org/apache/maven/plugins/jlink/JLinkExecutor.java

```diff
@@ -0,0 +1,171 @@
+{
+    private final String jLinkExec;
+
+    JLinkExecutor( Toolchain toolchain, Log log ) throws IOException
+    {
+        super( toolchain, log );
+        this.jLinkExec = getJLinkExecutable();
+    }
+
+    public File getJlinkExecutable()
+    {
+        return new File( this.jLinkExec );
+    }
+
+    @Override
+    public Optional<File> getJmodsFolder( /* nullable */ File sourceJdkModules )
+    {
+        // Really Hacky...do we have a better solution to find the jmods directory of the JDK?
+        File jLinkParent = getJlinkExecutable().getParentFile().getParentFile();
+        File jmodsFolder;
+        if ( sourceJdkModules != null && sourceJdkModules.isDirectory() )
+        {
+            jmodsFolder = new File( sourceJdkModules, JMODS );
+        }
+        else
+        {
+            jmodsFolder = new File( jLinkParent, JMODS );
+        }
+
+        getLog().debug( " Parent: " + jLinkParent.getAbsolutePath() );
+        getLog().debug( " jmodsFolder: " + jmodsFolder.getAbsolutePath() );
+
+        return Optional.of( jmodsFolder );
+    }
+
+    /**
+     * Execute JLink via any means.
+     *
+     * @return the exit code ({@code 0} on success).
+     */
+    @Override
+    public int executeJlink( File argsFile )
+    {
+        getLog().info( "Toolchain in maven-jlink-plugin: jlink [ " + this.jLinkExec + " ]" );
+
+        Commandline cmd = createJLinkCommandLine( argsFile );
+        cmd.setExecutable( this.jLinkExec );
+
+        throw new UnsupportedOperationException( "not implemented" );
+    }
+
+    private Commandline createJLinkCommandLine( File argsFile )
+    {
+        Commandline cmd = new Commandline();
+        cmd.createArg().setValue( '@' + argsFile.getAbsolutePath() );
+
+        return cmd;
+    }
+
+
+    protected final String getJLinkExecutable() throws IOException
+    {
+        String jLinkExecutable = null;
+        if ( getToolchain() != null )
+        {
+            jLinkExecutable = getToolchain().findTool( "jlink" );
+        }
+
+        // TODO: Check if there exist a more elegant way?
+        String jLinkCommand = "jlink" + ( SystemUtils.IS_OS_WINDOWS ? ".exe" : "" );
+
+        File jLinkExe;
+
+        if ( StringUtils.isNotEmpty( jLinkExecutable ) )
+        {
+            jLinkExe = new File( jLinkExecutable );
+
+            if ( jLinkExe.isDirectory() )
+            {
+                jLinkExe = new File( jLinkExe, jLinkCommand );
+            }
+
+            if ( SystemUtils.IS_OS_WINDOWS && jLinkExe.getName().indexOf( '.' ) < 0 )
+            {
+                jLinkExe = new File( jLinkExe.getPath() + ".exe" );
+            }
+
+            if ( !jLinkExe.isFile() )
+            {
+                throw new IOException( "The jlink executable '" + jLinkExe + "' doesn't exist or is not a file." );
+            }
+            return jLinkExe.getAbsolutePath();
+        }
+
+        // ----------------------------------------------------------------------
```

Comment:
I think everything below here can be replaced with a message that either a JDK 9+ as runtime is required or a toolchain